### ISSUE TITLE
False positive error due to unrelated assert in another function

### ISSUE THREAD

**Issue by chegoryu** [microsoft/pyright#9353]
**Describe the bug**

Pyright finds error in the next code:

```python
import pandas as pd

def f1():
    df = pd.DataFrame()
    assert df.columns[-1] == "abc"


def f2(time: pd.Series):
    pd.to_datetime(time).dt.date.unique()
```

```bash
$ PYRIGHT_PYTHON_FORCE_VERSION=1.1.386 /usr/bin/python3 -m pyright --pythonpath /usr/bin/python3 test.py
/home/echunaev/test.py
  /home/echunaev/test.py:9:34 - error: Cannot access attribute "unique" for class "ndarray[Unknown, Unknown]*"
    Attribute "unique" is unknown (reportAttributeAccessIssue)
1 error, 0 warnings, 0 informations
```

But if assert in `f1` is commented out, the error is not shown:

```python
import pandas as pd

def f1():
    _ = pd.DataFrame()
    # assert df.columns[-1] == "abc"


def f2(time: pd.Series):
    pd.to_datetime(time).dt.date.unique()
```

```bash
$ PYRIGHT_PYTHON_FORCE_VERSION=1.1.386 /usr/bin/python3 -m pyright --pythonpath /usr/bin/python3 test.py  
0 errors, 0 warnings, 0 informations
```

Python version and pandas version (I don't have a simple repro without pandas :():
```bash
$ python3 --version
Python 3.11.7
$ pip3 freeze | grep pandas
pandas==2.0.0rc1
pandas-stubs==1.5.3.230304
```

Also code above should be valid:

![Image](https://github.com/user-attachments/assets/05c94a9f-f346-4e3a-8b3a-9c497fdc64a9)

**VS Code extension or command-line**

Via command-line tool:
```
$ PYRIGHT_PYTHON_FORCE_VERSION=1.1.386 /usr/bin/python3 -m pyright --pythonpath /usr/bin/python3 test.py
```

No `pyproject.toml` is required (i.e. with default options).

**Comment by chegoryu** [comment]
Also updating `pandas-stub` to the actual version fixes the problem, but I guess that even with mismatched stubs the behaviour here is wrong.

**Comment by erictraut** [comment]
I'm not able to repro with the latest version of pandas-stubs. I'm also not able to install pandas-stubs version 1.5.3.230304; that version doesn't appear to be published in pypi. Can you tell me where I can get that version?

**Comment by chegoryu** [comment]
> I'm also not able to install pandas-stubs version 1.5.3.230304; that version doesn't appear to be published in pypi. Can you tell me where I can get that version?

Hmm, pip3 installs it:
```bash
$ pip3 install pandas-stubs==1.5.3.230304  
Defaulting to user installation because normal site-packages is not writeable
Collecting pandas-stubs==1.5.3.230304
  Downloading pandas_stubs-1.5.3.230304-py3-none-any.whl.metadata (9.4 kB)
Requirement already satisfied: types-pytz>=2022.1.1 in /usr/twix/python3.11/lib/python3.11/site-packages (from pandas-stubs==1.5.3.230304) (2022.7.1.2)
Downloading pandas_stubs-1.5.3.230304-py3-none-any.whl (149 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 149.8/149.8 kB 1.4 MB/s eta 0:00:00
Installing collected packages: pandas-stubs
  Attempting uninstall: pandas-stubs
    Found existing installation: pandas-stubs 2.2.3.241009
    Uninstalling pandas-stubs-2.2.3.241009:
      Successfully uninstalled pandas-stubs-2.2.3.241009
Successfully installed pandas-stubs-1.5.3.230304

[notice] A new release of pip is available: 24.0 -> 24.3.1
[notice] To update, run: pip install --upgrade pip

```

**Comment by erictraut** [comment]
I'm able to repro the problem, and I agree it looks like a bug. I'll investigate further.

**Comment by erictraut** [comment]
The issue is related to a TypeVar that has a circularity in its type definition.

The pandas stubs define a class `TimestampProperties` that derives from `DatetimeProperties` which derives from `_DatetimeNoTzProperties`. It uses a value-constrained type variable `_DTRoundingMethodReturnType`, and one of its constraints is `DatetimeIndex`. This class derives from `DatetimeIndexProperties` which derives from `_DatetimeNoTZProperties`. This creates a circularity in the dependency graph for these types. Depending on the order pyright enters this cycle, it may or may not encounter problems with partially-evaluated types. By adding or removing the `assert`, you are changing this order internally.

The [pyright documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=circular-references) contains some discussion of this type of circular reference.

I consider it a bug in the pandas stubs code that it includes such a circularity. And newer versions of the pandas stubs address this issue.

That said, I think that pyright can do a better job detecting and dealing with this particular form of circularity. I've updated the logic so the effects of this circularity are less disruptive. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.388.
